src/System.DirectoryServices.Protocols/tests/DirectoryServicesProtocolsTests.cs

```diff
@@ -0,0 +1,593 @@
+                    Assert.NotNull(sre);
+                    Assert.Equal("Modified Protocols Group 3", (string) sre.Attributes["description"][0]);
+
+                    DeleteAttribute(connection, dn, "description");
+                    sre = SearchOrganizationalUnit(connection, LdapConfiguration.Configuration.Domain, ouName);
+                    Assert.NotNull(sre);
+                    Assert.Null(sre.Attributes["description"]);
+                }
+                finally
+                {
+                    DeleteEntry(connection, dn);
+                }
+            }
+        }
+
+        [ConditionalFact(nameof(IsLdapConfigurationExist))]
+        public void TestNestedOUs()
+        {
+            using (LdapConnection connection = GetConnection())
+            {
+                string ouLevel1Name = "ProtocolsGroup4-1";
+                string dnLevel1 = "ou=" + ouLevel1Name;
+                string ouLevel2Name = "ProtocolsGroup4-2";
+                string dnLevel2 = "ou=" + ouLevel2Name+ "," + dnLevel1;
+
+                DeleteEntry(connection, dnLevel2);
+                DeleteEntry(connection, dnLevel1);
+
+                try
+                {
+                    AddOrganizationalUnit(connection, dnLevel1);
+                    SearchResultEntry sre = SearchOrganizationalUnit(connection, LdapConfiguration.Configuration.Domain, ouLevel1Name);
+                    Assert.NotNull(sre);
+
+                    AddOrganizationalUnit(connection, dnLevel2);
+                    sre = SearchOrganizationalUnit(connection, dnLevel1 + "," + LdapConfiguration.Configuration.Domain, ouLevel2Name);
+                    Assert.NotNull(sre);
+                }
+                finally
+                {
+                    DeleteEntry(connection, dnLevel2);
+                    DeleteEntry(connection, dnLevel1);
+                }
+            }
+        }
+
+        [ConditionalFact(nameof(IsLdapConfigurationExist))]
+        public void TestAddUser()
+        {
+            using (LdapConnection connection = GetConnection())
+            {
+                string ouName = "ProtocolsGroup5";
+                string dn = "ou=" + ouName;
+                string user1Dn = "cn=protocolUser1" + "," + dn;
+                string user2Dn = "cn=protocolUser2" + "," + dn;
+
+                DeleteEntry(connection, user1Dn);
+                DeleteEntry(connection, user2Dn);
+                DeleteEntry(connection, dn);
+
+                try
+                {
+                    AddOrganizationalUnit(connection, dn);
+                    SearchResultEntry sre = SearchOrganizationalUnit(connection, LdapConfiguration.Configuration.Domain, ouName);
+                    Assert.NotNull(sre);
+
+                    AddOrganizationalRole(connection, user1Dn);
+                    AddOrganizationalRole(connection, user2Dn);
+
+                    string usersRoot = dn + "," + LdapConfiguration.Configuration.Domain;
```

Comment:
We should setup an LDS instance and make sure this works.  LDS has some specific requirements on how paths are used.

Reply:
could you please explain more how I can do that? or if you have a pointer to some example that will be great

Reply:
do you mean here connecting to the domain which the local machine is connected to?

Reply:
Is what you mean here is doing https://technet.microsoft.com/en-us/library/cc725619(v=ws.10).aspx?